### ISSUE TITLE
Numerical Instability for apply_decay_mask

### ISSUE THREAD

**Issue by boxin-wbx** [timoschick/self-debiasing#3]
`apply_dacy_mask` convert `logits` to `probability` via softmax. In [generation.py](https://github.com/timoschick/self-debiasing/blob/762f1811cc361014e92f0b6f7ffdd7f3e8d3200d/generation.py#L46), the probability is then converted back to logits via `torch.log`, which may cause numerical instability. In my case, during the ppl evaluation, I encountered some probabilities became 0 and the logits became -inf, which makes the ppl extremely large.

To solve the issue, I wrote an equivalent version below:
```python
def apply_decay_mask_logits(args, logits: torch.Tensor, decay_mask: torch.Tensor) -> torch.Tensor:
    """Applies exponential decay to a tensor of logits"""
    decay_mask = torch.exp(- decay_mask * args.decay_constant)
    decay_mask = torch.max(decay_mask, torch.tensor([args.epsilon], device=decay_mask.device))
    log_decay_mask = torch.log(decay_mask)
    logits += log_decay_mask
    return logits
```

Please advise. If it looks good to you, I can submit a pull request :)

**Comment by timoschick** [comment]
Hi @boxin-wbx, just out of curiosity, did you encounter those issues when trying to reproduce our results (i.e., using our exact set of parameters) or using some different setup?

Regardless, your proposed change is definitely worth implementing, so I'd be happy if you could submit a pull request :)

**Comment by boxin-wbx** [comment]
Thanks for the review. I did not try to reproduce your results. I tried to adapt your code to Megatron LM, and encountered the numerical instability issue. My current approach directly works on the logits space and converts the decay mask to the log space to avoid the issue.
